Multiple unit test failures with postgresql

FAILED tests/test_errors.py::test_empty_results - firepit.exceptions.UnknownViewname: relation "my_findings" does not exist
FAILED tests/test_storage.py::test_ops[url-value-MATCHES-^.*example.com/page/1[0-9]*$-http://www26.example.com/page/176-http://www67.example.com/page/264] - AssertionErro...
FAILED tests/test_storage.py::test_grouping - firepit.exceptions.IncompatibleType
FAILED tests/test_storage.py::test_extract - firepit.exceptions.IncompatibleType
FAILED tests/test_storage.py::test_reassign - KeyError: 'x_enrich'
FAILED tests/test_storage.py::test_appdata - TypeError: the JSON object must be str, bytes or bytearray, not NoneType
FAILED tests/test_storage.py::test_sort_same_name - psycopg2.errors.DependentObjectsStillExist: cannot drop view gmirloov because other objects depend on it

--- a/docs/usage.rst
+++ b/docs/usage.rst
@@ -44,7 +44,9 @@ data organized.
       Columnar storage for STIX observations
 
     Options:
-      --help  Show this message and exit.
+      --dbname TEXT   Path/name of database  [default: stix.db]
+      --session TEXT  Session ID to data separation  [default: test-id]
+      --help          Show this message and exit.
 
     Commands:
       assign       Perform an operation on a column and name the result
@@ -61,7 +63,7 @@ data organized.
       merge        Merge 2 or more views into a new view
       reassign     Update/replace STIX observation data in SQL
       remove       Remove a view
-      rename       Remove a view
+      rename       Rename a view
       schema       Get the schema of a view/table
       set-appdata  Set the app-specific data for a view
       sql          Run a SQL statement on the database [DANGEROUS!]
@@ -81,6 +83,4 @@ data organized.
       FILENAMES...  STIX bundle files of query results  [required]
 
     Options:
-      --dbname TEXT   Path/name of database  [default: stix.db]
-      --session TEXT  Session ID to data separation  [default: test-id]
-      --help          Show this message and exit.
+      --help  Show this message and exit.

--- a/firepit/__init__.py
+++ b/firepit/__init__.py
@@ -2,7 +2,7 @@
 
 __author__ = """IBM Security"""
 __email__ = 'pcoccoli@us.ibm.com'
-__version__ = '1.0.14'
+__version__ = '1.0.15'
 
 
 from importlib import import_module

--- a/firepit/cli.py
+++ b/firepit/cli.py
@@ -19,14 +19,6 @@ from firepit import get_storage
 logging.basicConfig(level=os.environ.get("LOGLEVEL", "INFO"))
 
 
-def defdb():
-    return os.getenv('FIREPITDB', 'stix.db')
-
-
-def defid():
-    return os.getenv('FIREPITID', 'test-id')
-
-
 app = typer.Typer(
     name="firepit",
     help="Columnar storage for STIX observations",
@@ -34,15 +26,30 @@ app = typer.Typer(
 )
 
 
+state = {
+    'dbname': os.getenv('FIREPITDB', 'stix.db'),
+    'session': os.getenv('FIREPITID', 'test-id'),
+}
+
+
+@app.callback()
+def main(
+    dbname: str = typer.Option(None, help="Path/name of database"),
+    session: str = typer.Option(state['session'], help="Session ID to data separation"),
+):
+    if dbname:
+        state['dbname'] = dbname
+    if session:
+        state['session'] = session
+
+
 @app.command()
 def cache(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     query_id: str = typer.Argument(..., help="An identifier for this set of data"),
     filenames: List[str] = typer.Argument(..., help="STIX bundle files of query results"),
 ):
     """Cache STIX observation data in SQL"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     if isinstance(filenames, tuple):
         filenames = list(filenames)
     db.cache(query_id, filenames)
@@ -50,36 +57,30 @@ def cache(
 
 @app.command()
 def extract(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="Name for this new view"),
     sco_type: str = typer.Argument(..., help="SCO type to extract"),
     query_id: str = typer.Argument(..., help="Identifier for cached data to extract from"),
     pattern: str = typer.Argument(..., help="STIX pattern to filter cached data"),
 ):
     """Create a view of a subset of cached data"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     db.extract(name, sco_type, query_id, pattern)
 
 
 @app.command()
 def filter(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="Name for this new view"),
     sco_type: str = typer.Argument(..., help="SCO type to extract"),
     source: str = typer.Argument(..., help="Source view"),
     pattern: str = typer.Argument(..., help="STIX pattern to filter cached data"),
 ):
     """Create a filtered view of a subset of cached data"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     db.filter(name, sco_type, source, pattern)
 
 
 @app.command()
 def assign(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="Name for this new view"),
     view: str = typer.Argument(..., help="View name to operate on"),
     op: str = typer.Option(..., help="Operation to perform (sort, group, etc.)"),
@@ -88,15 +89,13 @@ def assign(
     limit: int = typer.Option(None, help="Max number of rows to return"),
 ):
     """Perform an operation on a column and name the result"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     asc = not desc
     db.assign(name, view, op, by, asc, limit)
 
 
 @app.command()
 def join(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="Name for this new view"),
     left_view: str = typer.Argument(..., help="Left view name to join"),
     left_on: str = typer.Argument(..., help="Column from left view to join on"),
@@ -104,21 +103,19 @@ def join(
     right_on: str = typer.Argument(..., help="Column from right view to join on"),
 ):
     """Join two views"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     db.join(name, left_view, left_on, right_view, right_on)
 
 
 @app.command()
 def lookup(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="View name to look up"),
     limit: int = typer.Option(None, help="Max number of rows to return"),
     offset: int = typer.Option(0, help="Number of rows to skip"),
     format: str = typer.Option('table', help="Output format [table, json]"),
 ):
     """Retrieve a view"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     rows = db.lookup(name, limit=limit, offset=offset)
     if format == 'json':
         print(json.dumps(rows, separators=[',', ':']))
@@ -128,37 +125,29 @@ def lookup(
 
 @app.command()
 def values(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     path: str = typer.Argument(..., help="STIX object path to retrieve from view"),
     name: str = typer.Argument(..., help="View name to look up"),
 ):
     """Retrieve the values of a STIX object path from a view"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     rows = db.values(path, name)
     for row in rows:
         print(row)
 
 
 @app.command()
-def tables(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
-):
+def tables():
     """Get all view/table names"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     rows = db.tables()
     for row in rows:
         print(row)
 
 
 @app.command()
-def views(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
-):
+def views():
     """Get all view names"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     rows = db.views()
     for row in rows:
         print(row)
@@ -166,13 +155,11 @@ def views(
 
 @app.command()
 def viewdata(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     views: List[str] = typer.Argument(None, help="Views to merge"),
     format: str = typer.Option('table', help="Output format [table, json]"),
 ):
     """Get view data for views [default is all views]"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     rows = db.get_view_data(views)
     if format == 'json':
         print(json.dumps(rows, separators=[',', ':']))
@@ -182,12 +169,10 @@ def viewdata(
 
 @app.command()
 def columns(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="View name to look up"),
 ):
     """Get the columns names of a view/table"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     rows = db.columns(name)
     for row in rows:
         print(row)
@@ -195,56 +180,46 @@ def columns(
 
 @app.command()
 def type(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="View name to look up"),
 ):
     """Get the SCO type of a view/table"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     print(db.table_type(name))
 
 
 @app.command()
 def schema(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="View name to look up"),
 ):
     """Get the schema of a view/table"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     rows = db.schema(name)
     print(tabulate(rows, headers='keys'))
 
 
 @app.command()
 def count(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="View name to look up"),
 ):
     """Get the count of rows in a view/table"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     print(db.count(name))
 
 
 @app.command()
 def delete(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
 ):
     """Delete STIX observation data in SQL"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     db.delete()
 
 
 @app.command()
 def sql(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     statement: str = typer.Argument(..., help="View name to look up"),
 ):
     """Run a SQL statement on the database [DANGEROUS!]"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     rows = db._execute(statement)
     if rows:
         print(tabulate(rows, headers='keys'))
@@ -252,31 +227,25 @@ def sql(
 
 @app.command()
 def set_appdata(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="View name"),
     data: str = typer.Argument(..., help="Data (string)"),
 ):
     """Set the app-specific data for a view"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     db.set_appdata(name, data)
 
 
 @app.command()
 def get_appdata(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="View name"),
 ):
     """Get the app-specific data for a view"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     print(db.get_appdata(name))
 
 
 @app.command()
 def load(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="Name for this new view"),
     sco_type: str = typer.Option(None, help="SCO type of data to load"),
     query_id: str = typer.Option(None, help="An identifier for this set of data"),
@@ -284,7 +253,7 @@ def load(
     filename: str = typer.Argument(..., help="Data file to load (JSON only)"),
 ):
     """Cache STIX observation data in SQL"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     try:
         with open(filename, 'r') as fp:
             data = json.load(fp)
@@ -296,13 +265,11 @@ def load(
 
 @app.command()
 def reassign(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="Name for this new view"),
     filename: str = typer.Argument(..., help="Data file to load (JSON only)"),
 ):
     """Update/replace STIX observation data in SQL"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     with open(filename, 'r') as fp:
         data = json.load(fp)
     db.reassign(name, data)
@@ -310,36 +277,30 @@ def reassign(
 
 @app.command()
 def merge(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="Name for this new view"),
     views: List[str] = typer.Argument(..., help="Views to merge"),
 ):
     """Merge 2 or more views into a new view"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     db.merge(name, views)
 
 
 @app.command()
 def remove(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     name: str = typer.Argument(..., help="Name of view to remove"),
 ):
     """Remove a view"""
-    db = get_storage(dbname, session)
+    db = get_storage(state['dbname'], state['session'])
     db.remove_view(name)
 
 
 @app.command()
 def rename(
-    dbname: str = typer.Option(defdb(), help="Path/name of database"),
-    session: str = typer.Option(defid(), help="Session ID to data separation"),
     oldname: str = typer.Argument(..., help="Name of view to rename"),
     newname: str = typer.Argument(..., help="New name of view to rename"),
 ):
-    """Remove a view"""
-    db = get_storage(dbname, session)
+    """Rename a view"""
+    db = get_storage(state['dbname'], state['session'])
     db.rename_view(oldname, newname)
 
 

--- a/firepit/pgstorage.py
+++ b/firepit/pgstorage.py
@@ -1,8 +1,6 @@
 import logging
 import os
-import random
 import re
-import string
 
 import orjson
 import psycopg2
@@ -65,7 +63,7 @@ class PgStorage(SqlStorage):
         try:
             self._execute('''CREATE FUNCTION match(pattern TEXT, value TEXT)
                 RETURNS boolean AS $$
-                    SELECT regexp_match(value, pattern) != '{}'
+                    SELECT regexp_match(value, pattern) IS NOT NULL;
             $$ LANGUAGE SQL;''', cursor=cursor)
         except psycopg2.errors.DuplicateFunction:
             self.connection.rollback()
@@ -144,34 +142,23 @@ class PgStorage(SqlStorage):
         validate_name(viewname)
         if not cursor:
             cursor = self._execute('BEGIN;')
-        tmp = None
         if not deps:
             deps = []
         elif viewname in deps:
-            # Rename old view to random var
-            tmp = ''.join(random.choice(string.ascii_lowercase)
-                          for x in range(8))
-            self._execute(f'DROP VIEW IF EXISTS "{tmp}"', cursor)
+            # Get the query that makes up the current view
             slct = self._get_view_def(viewname)
-            self._create_view(tmp, slct, sco_type, cursor=cursor)
-            select = re.sub(f'"{viewname}"', tmp, select)
-        # Check if deps exist
-        tables = self.tables()
-        for dep in deps:
-            if dep not in tables:
-                break
-        else:
-            try:
-                self._execute(f'CREATE OR REPLACE VIEW "{viewname}" AS {select}', cursor)
-            except psycopg2.errors.UndefinedTable:
-                self.connection.rollback()
-                cursor = self._execute('BEGIN;')
-                self._create_empty_view(viewname, cursor)
-            except psycopg2.errors.InvalidTableDefinition:
-                self.connection.rollback()
-                raise IncompatibleType
-        if tmp:
-            self._execute(f'DROP VIEW IF EXISTS "{tmp}"', cursor)
+            self._execute(f'DROP VIEW IF EXISTS "{viewname}"', cursor)
+            # Swap out the viewname for its definition
+            select = re.sub(f'"{viewname}"', f'({slct}) AS tmp', select)
+        try:
+            self._execute(f'CREATE OR REPLACE VIEW "{viewname}" AS {select}', cursor)
+        except psycopg2.errors.UndefinedTable:
+            self.connection.rollback()
+            cursor = self._execute('BEGIN;')
+            self._create_empty_view(viewname, cursor)
+        except psycopg2.errors.InvalidTableDefinition:
+            self.connection.rollback()
+            raise IncompatibleType
         self._new_name(cursor, viewname, sco_type)
         return cursor
 
@@ -244,11 +231,8 @@ class PgStorage(SqlStorage):
             query_values.append(obj['id'])
             query_values.append(query_id)
             for c in colnames:
-                if obj.get(c):
-                    value = obj[c]
-                    values.append(str(orjson.dumps(value), 'utf-8') if isinstance(value, list) else value)
-                else:
-                    values.append(None)
+                value = obj.get(c, None)
+                values.append(str(orjson.dumps(value), 'utf-8') if isinstance(value, list) else value)
         logger.debug('_upsert: "%s"', stmt)
         cursor.execute(stmt, values)
 

--- a/firepit/sqlstorage.py
+++ b/firepit/sqlstorage.py
@@ -107,7 +107,7 @@ class SqlStorage:
         if cols != "*":
             cols = ", ".join([f'"{col}"' if not col.startswith("'") else col for col in cols])
 
-        stmt = f'SELECT {cols} from "{tvname}"'
+        stmt = f'SELECT {cols} FROM "{tvname}"'
         if where:
             stmt += f' WHERE {where}'
         if groupby:

--- a/setup.cfg
+++ b/setup.cfg
@@ -19,7 +19,3 @@ exclude = docs
 
 [aliases]
 test = pytest
-
-[tool:pytest]
-collect_ignore = ['setup.py']
-

--- a/setup.cfg
+++ b/setup.cfg
@@ -1,5 +1,5 @@
 [bumpversion]
-current_version = 1.0.14
+current_version = 1.0.15
 commit = True
 tag = True
 
@@ -19,3 +19,4 @@ exclude = docs
 
 [aliases]
 test = pytest
+

--- a/setup.py
+++ b/setup.py
@@ -55,6 +55,6 @@ setup(
     test_suite='tests',
     tests_require=test_requirements,
     url='https://github.com/IBM/firepit',
-    version='1.0.14',
+    version='1.0.15',
     zip_safe=False,
 )

--- a/tests/ccoe_investigator_demo.json
+++ b/tests/ccoe_investigator_demo.json
@@ -103305,14 +103305,14 @@
 		},
 		{
 			"type": "identity",
-			"id": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"id": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"name": "shifter_proxy",
 			"identity_class": "events"
 		},
 		{
 			"id": "observed-data--4054384e-5f99-4bf0-acd4-8f746cbedcdd",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:34.950Z",
 			"modified": "2021-01-14T21:31:34.950Z",
 			"objects": {
@@ -103487,7 +103487,7 @@
 		{
 			"id": "observed-data--ab5c98db-16b2-425b-a0ba-acd95715aa56",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:34.978Z",
 			"modified": "2021-01-14T21:31:34.978Z",
 			"objects": {
@@ -103621,7 +103621,7 @@
 		{
 			"id": "observed-data--a1654786-bded-4639-ac2d-3b13c7fc3d32",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.062Z",
 			"modified": "2021-01-14T21:31:35.062Z",
 			"objects": {
@@ -103755,7 +103755,7 @@
 		{
 			"id": "observed-data--9c3dbfb1-5ee6-43a3-98df-b87a379de72d",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.092Z",
 			"modified": "2021-01-14T21:31:35.093Z",
 			"objects": {
@@ -103906,7 +103906,7 @@
 		{
 			"id": "observed-data--1c0fdce1-ce29-461a-bf99-e79fa18dce85",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.139Z",
 			"modified": "2021-01-14T21:31:35.139Z",
 			"objects": {
@@ -104040,7 +104040,7 @@
 		{
 			"id": "observed-data--f0a3e1b0-507a-41ea-a204-51e99eb97106",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.171Z",
 			"modified": "2021-01-14T21:31:35.171Z",
 			"objects": {
@@ -104215,7 +104215,7 @@
 		{
 			"id": "observed-data--6acc3bc0-3f97-4083-aa48-95a1722cb63a",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.208Z",
 			"modified": "2021-01-14T21:31:35.208Z",
 			"objects": {
@@ -104351,7 +104351,7 @@
 		{
 			"id": "observed-data--0d2f99e9-3360-48bb-9e90-3f84f867e268",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.240Z",
 			"modified": "2021-01-14T21:31:35.240Z",
 			"objects": {
@@ -104526,7 +104526,7 @@
 		{
 			"id": "observed-data--24198e46-449d-4058-b745-88aadee0f7d7",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.253Z",
 			"modified": "2021-01-14T21:31:35.253Z",
 			"objects": {
@@ -104656,7 +104656,7 @@
 		{
 			"id": "observed-data--87b8e880-41c2-483f-be92-58568caec29d",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.297Z",
 			"modified": "2021-01-14T21:31:35.297Z",
 			"objects": {
@@ -104786,7 +104786,7 @@
 		{
 			"id": "observed-data--3bac6f29-e474-4872-bcc8-cedcc51d22b2",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.355Z",
 			"modified": "2021-01-14T21:31:35.355Z",
 			"objects": {
@@ -104916,7 +104916,7 @@
 		{
 			"id": "observed-data--eb467ea5-ef3f-491d-8908-fdefbf3f6f9c",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.413Z",
 			"modified": "2021-01-14T21:31:35.413Z",
 			"objects": {
@@ -105062,7 +105062,7 @@
 		{
 			"id": "observed-data--b32e7ac5-c301-4e21-a9a5-affd37c80578",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.460Z",
 			"modified": "2021-01-14T21:31:35.460Z",
 			"objects": {
@@ -105208,7 +105208,7 @@
 		{
 			"id": "observed-data--778d7344-09e0-405b-ad2a-a41e131e7410",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.486Z",
 			"modified": "2021-01-14T21:31:35.486Z",
 			"objects": {
@@ -105344,7 +105344,7 @@
 		{
 			"id": "observed-data--6481c198-8767-48ab-aa4c-ede25666ba9c",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.534Z",
 			"modified": "2021-01-14T21:31:35.534Z",
 			"objects": {
@@ -105519,7 +105519,7 @@
 		{
 			"id": "observed-data--31f7efa6-0ace-4fcd-b2b2-860d7b72224e",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.563Z",
 			"modified": "2021-01-14T21:31:35.563Z",
 			"objects": {
@@ -105714,7 +105714,7 @@
 		{
 			"id": "observed-data--5989d31e-dc15-4868-90c1-53d1887aaf94",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.595Z",
 			"modified": "2021-01-14T21:31:35.595Z",
 			"objects": {
@@ -105865,7 +105865,7 @@
 		{
 			"id": "observed-data--6b7a9fc2-2b5c-4f88-be74-e87d3a07379e",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.667Z",
 			"modified": "2021-01-14T21:31:35.667Z",
 			"objects": {
@@ -106051,7 +106051,7 @@
 		{
 			"id": "observed-data--d11332c3-af8e-411f-939f-549b8bf71397",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.697Z",
 			"modified": "2021-01-14T21:31:35.697Z",
 			"objects": {
@@ -106235,7 +106235,7 @@
 		{
 			"id": "observed-data--6d481df9-3bbe-4a4d-bb4f-3c6613cd6bb8",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.725Z",
 			"modified": "2021-01-14T21:31:35.725Z",
 			"objects": {
@@ -106386,7 +106386,7 @@
 		{
 			"id": "observed-data--38372c15-b155-4750-94a7-377dcd9920ac",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.764Z",
 			"modified": "2021-01-14T21:31:35.764Z",
 			"objects": {
@@ -106532,7 +106532,7 @@
 		{
 			"id": "observed-data--a815751a-d8da-451c-8fef-a3a74dbbf47d",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.831Z",
 			"modified": "2021-01-14T21:31:35.831Z",
 			"objects": {
@@ -106663,7 +106663,7 @@
 		{
 			"id": "observed-data--e9771be6-df70-41f4-b027-e6082f0a1eeb",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.889Z",
 			"modified": "2021-01-14T21:31:35.889Z",
 			"objects": {
@@ -106838,7 +106838,7 @@
 		{
 			"id": "observed-data--3b448df7-7f27-4683-90da-89f1f9a81e18",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.900Z",
 			"modified": "2021-01-14T21:31:35.900Z",
 			"objects": {
@@ -107013,7 +107013,7 @@
 		{
 			"id": "observed-data--a2b88648-52db-4552-8d66-11451258e85e",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.910Z",
 			"modified": "2021-01-14T21:31:35.910Z",
 			"objects": {
@@ -107200,7 +107200,7 @@
 		{
 			"id": "observed-data--60733f6e-deae-4e38-86ac-a734d4a3e9a1",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.919Z",
 			"modified": "2021-01-14T21:31:35.919Z",
 			"objects": {
@@ -107351,7 +107351,7 @@
 		{
 			"id": "observed-data--00854ed2-1952-4118-967e-bc858aa2202f",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.931Z",
 			"modified": "2021-01-14T21:31:35.931Z",
 			"objects": {
@@ -107502,7 +107502,7 @@
 		{
 			"id": "observed-data--60df2df3-0d9d-4957-a002-94eb71f3afd4",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.944Z",
 			"modified": "2021-01-14T21:31:35.944Z",
 			"objects": {
@@ -107638,7 +107638,7 @@
 		{
 			"id": "observed-data--b0bc2fd5-2a8a-485f-bc76-d5f5d9f84de0",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.958Z",
 			"modified": "2021-01-14T21:31:35.958Z",
 			"objects": {
@@ -107825,7 +107825,7 @@
 		{
 			"id": "observed-data--f784555c-4b7c-4b50-8110-8fe0925ce743",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.966Z",
 			"modified": "2021-01-14T21:31:35.966Z",
 			"objects": {
@@ -107961,7 +107961,7 @@
 		{
 			"id": "observed-data--7cb4d458-c081-41d6-8cbc-a1bc81999f42",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.980Z",
 			"modified": "2021-01-14T21:31:35.980Z",
 			"objects": {
@@ -108097,7 +108097,7 @@
 		{
 			"id": "observed-data--d820f56f-5ad8-4e7e-a11c-66c999b38d5a",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:35.994Z",
 			"modified": "2021-01-14T21:31:35.994Z",
 			"objects": {
@@ -108272,7 +108272,7 @@
 		{
 			"id": "observed-data--a9445c3c-12ed-4194-87d1-4a0580b742d3",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:36.005Z",
 			"modified": "2021-01-14T21:31:36.005Z",
 			"objects": {
@@ -108447,7 +108447,7 @@
 		{
 			"id": "observed-data--fe63ebf6-a76f-4679-af01-855a244fac76",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:36.015Z",
 			"modified": "2021-01-14T21:31:36.015Z",
 			"objects": {
@@ -108572,7 +108572,7 @@
 		{
 			"id": "observed-data--1e36a13c-832f-461b-a96b-66ae2cf033b4",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:36.029Z",
 			"modified": "2021-01-14T21:31:36.029Z",
 			"objects": {
@@ -108697,7 +108697,7 @@
 		{
 			"id": "observed-data--b5adc875-710c-446a-8a3e-605f49cac2a1",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:36.043Z",
 			"modified": "2021-01-14T21:31:36.043Z",
 			"objects": {
@@ -108822,7 +108822,7 @@
 		{
 			"id": "observed-data--9978542a-9969-4f4f-90f1-77615790ef86",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:36.057Z",
 			"modified": "2021-01-14T21:31:36.057Z",
 			"objects": {
@@ -108958,7 +108958,7 @@
 		{
 			"id": "observed-data--d1b81d23-a3af-45d7-903a-1ae2e703f9c2",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:36.071Z",
 			"modified": "2021-01-14T21:31:36.071Z",
 			"objects": {
@@ -109083,7 +109083,7 @@
 		{
 			"id": "observed-data--5c63a5f0-7e26-4c6e-bbad-b7e0bbc497b0",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:36.085Z",
 			"modified": "2021-01-14T21:31:36.085Z",
 			"objects": {
@@ -109219,7 +109219,7 @@
 		{
 			"id": "observed-data--17511d66-f52d-4361-a1ad-46a95b784b95",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:36.098Z",
 			"modified": "2021-01-14T21:31:36.099Z",
 			"objects": {
@@ -109344,7 +109344,7 @@
 		{
 			"id": "observed-data--e753a510-7ddd-46fe-b02c-7ce9d6615070",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:36.113Z",
 			"modified": "2021-01-14T21:31:36.113Z",
 			"objects": {
@@ -109469,7 +109469,7 @@
 		{
 			"id": "observed-data--b25c90db-5828-40d9-86ef-adb41cc29c2e",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:36.127Z",
 			"modified": "2021-01-14T21:31:36.127Z",
 			"objects": {
@@ -109605,7 +109605,7 @@
 		{
 			"id": "observed-data--e7fb9739-0279-4b0e-be23-acab160ab778",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:36.141Z",
 			"modified": "2021-01-14T21:31:36.141Z",
 			"objects": {
@@ -109780,7 +109780,7 @@
 		{
 			"id": "observed-data--fdd3287c-9866-4bec-8c13-ab43498ac335",
 			"type": "observed-data",
-			"created_by_ref": "identity--f431f809-377b-45e0-aa1c-6a4751cae5ff",
+			"created_by_ref": "identity--ba4e0b74-fa91-4e7b-abf4-ca5ba68ca9ab",
 			"created": "2021-01-14T21:31:36.150Z",
 			"modified": "2021-01-14T21:31:36.150Z",
 			"objects": {

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -11,6 +11,12 @@ def fake_bundle_file():
 
 
 @pytest.fixture
+def fake_bundle_file_2():
+    cwd = os.path.dirname(os.path.abspath(__file__))
+    return os.path.join(cwd, 'test_bundle_2.json')
+
+
+@pytest.fixture
 def ccoe_bundle():
     cwd = os.path.dirname(os.path.abspath(__file__))
     return os.path.join(cwd, 'ccoe_investigator_demo.json')

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1,6 +1,8 @@
 import os
 import pytest
 
+collect_ignore = ['setup.py']
+
 
 @pytest.fixture
 def fake_bundle_file():

--- a/tests/helpers.py
+++ b/tests/helpers.py
@@ -1,5 +1,13 @@
-from firepit.sqlitestorage import SQLiteStorage
+import os
+from firepit import get_storage
 
+def tmp_storage(tmpdir, clear=True):
+    dbname = os.getenv('FIREPITDB', str(tmpdir.join('test.db')))
+    session = os.getenv('FIREPITID', 'test-session')
 
-def tmp_storage(tmpdir):
-    return SQLiteStorage(str(tmpdir.join('test.db')))
+    if clear:
+        # Clear out previous test session
+        store = get_storage(dbname, session)
+        store.delete()
+
+    return get_storage(dbname, session)

--- a/tests/test_bundle_2.json
+++ b/tests/test_bundle_2.json
@@ -1,0 +1,405 @@
+{
+  "type": "bundle",
+  "id": "bundle--4ce87790-ac67-40c5-8ec5-70755b0e28cd",
+  "spec_version": "2.0",
+  "objects": [
+    {
+      "type": "identity",
+      "id": "identity--89c5ac33-4656-4aa3-be68-f78879e5e273",
+      "created": "2020-06-30T19:31:23.304Z",
+      "modified": "2021-07-14T19:45:16.866Z",
+      "name": "stix-generator.py",
+      "identity_class": "program"
+    },
+    {
+      "type": "observed-data",
+      "id": "observed-data--2171d844-d635-4f03-91cc-0a36f1caf3b6",
+      "created_by_ref": "identity--89c5ac33-4656-4aa3-be68-f78879e5e273",
+      "created": "2020-06-30T19:31:23.651Z",
+      "modified": "2021-07-14T19:45:16.867Z",
+      "first_observed": "2020-06-30T19:25:09.447726Z",
+      "last_observed": "2020-06-30T19:28:49.692424Z",
+      "number_observed": 1,
+      "objects": {
+        "0": {
+          "type": "ipv4-addr",
+          "value": "192.168.212.97"
+        },
+        "1": {
+          "type": "ipv4-addr",
+          "value": "10.0.0.141"
+        },
+        "2": {
+          "type": "network-traffic",
+          "start": "2020-06-30T19:25:09.447726Z",
+          "end": "2020-06-30T19:28:49.692424Z",
+          "src_ref": "0",
+          "dst_ref": "1",
+          "src_port": 0,
+          "dst_port": 0,
+          "protocols": [
+            "ip",
+            "tcp"
+          ]
+        },
+        "3": {
+          "type": "user-account",
+          "user_id": "1008",
+          "account_login": "henry"
+        }
+      }
+    },
+    {
+      "type": "observed-data",
+      "id": "observed-data--827948c1-c4e9-49a5-85c7-5c25d375a890",
+      "created_by_ref": "identity--89c5ac33-4656-4aa3-be68-f78879e5e273",
+      "created": "2020-06-30T19:31:23.412Z",
+      "modified": "2021-07-14T19:45:16.867Z",
+      "first_observed": "2020-06-30T19:25:10.723267Z",
+      "last_observed": "2020-06-30T19:25:34.983662Z",
+      "number_observed": 1,
+      "objects": {
+        "0": {
+          "type": "ipv4-addr",
+          "value": "192.168.121.121"
+        },
+        "1": {
+          "type": "ipv4-addr",
+          "value": "10.0.0.134"
+        },
+        "2": {
+          "type": "network-traffic",
+          "start": "2020-06-30T19:25:10.723267Z",
+          "end": "2020-06-30T19:25:34.983662Z",
+          "src_ref": "0",
+          "dst_ref": "1",
+          "src_port": 49434,
+          "dst_port": 3128,
+          "protocols": [
+            "ip",
+            "tcp",
+            "http"
+          ]
+        },
+        "3": {
+          "type": "url",
+          "value": "http://www27.example.com/page/64"
+        },
+        "4": {
+          "type": "user-account",
+          "user_id": "1002",
+          "account_login": "bob"
+        }
+      }
+    },
+    {
+      "type": "observed-data",
+      "id": "observed-data--38b48bb3-d2cc-4a7f-8800-e8c643f295b7",
+      "created_by_ref": "identity--89c5ac33-4656-4aa3-be68-f78879e5e273",
+      "created": "2020-06-30T19:31:23.510Z",
+      "modified": "2021-07-14T19:45:16.867Z",
+      "first_observed": "2020-06-30T19:25:13.068734Z",
+      "last_observed": "2020-06-30T19:26:03.959427Z",
+      "number_observed": 1,
+      "objects": {
+        "0": {
+          "type": "ipv4-addr",
+          "value": "192.168.152.147"
+        },
+        "1": {
+          "type": "ipv4-addr",
+          "value": "10.0.0.214"
+        },
+        "2": {
+          "type": "network-traffic",
+          "start": "2020-06-30T19:25:13.068734Z",
+          "end": "2020-06-30T19:26:03.959427Z",
+          "src_ref": "0",
+          "dst_ref": "1",
+          "src_port": 53557,
+          "dst_port": 8080,
+          "protocols": [
+            "ip",
+            "tcp",
+            "http"
+          ]
+        },
+        "3": {
+          "type": "url",
+          "value": "http://www77.example.com/page/1"
+        },
+        "4": {
+          "type": "user-account",
+          "user_id": "1010",
+          "account_login": "juan"
+        }
+      }
+    },
+    {
+      "type": "observed-data",
+      "id": "observed-data--17dd077b-9ccb-4ab2-9d7e-be91f78f7d3f",
+      "created_by_ref": "identity--89c5ac33-4656-4aa3-be68-f78879e5e273",
+      "created": "2020-06-30T19:31:23.464Z",
+      "modified": "2021-07-14T19:45:16.867Z",
+      "first_observed": "2020-06-30T19:25:20.119776Z",
+      "last_observed": "2020-06-30T19:26:40.634223Z",
+      "number_observed": 1,
+      "objects": {
+        "0": {
+          "type": "ipv4-addr",
+          "value": "192.168.52.133"
+        },
+        "1": {
+          "type": "ipv4-addr",
+          "value": "10.0.0.201"
+        },
+        "2": {
+          "type": "network-traffic",
+          "start": "2020-06-30T19:25:20.119776Z",
+          "end": "2020-06-30T19:26:40.634223Z",
+          "src_ref": "0",
+          "dst_ref": "1",
+          "src_port": 61700,
+          "dst_port": 514,
+          "protocols": [
+            "ip",
+            "tcp"
+          ]
+        },
+        "3": {
+          "type": "user-account",
+          "user_id": "1005",
+          "account_login": "eve"
+        }
+      }
+    },
+    {
+      "type": "observed-data",
+      "id": "observed-data--4406e0b8-b0f7-4c0f-8085-0198efde7b12",
+      "created_by_ref": "identity--89c5ac33-4656-4aa3-be68-f78879e5e273",
+      "created": "2020-06-30T19:31:23.660Z",
+      "modified": "2021-07-14T19:45:16.867Z",
+      "first_observed": "2020-06-30T19:25:23.518799Z",
+      "last_observed": "2020-06-30T19:27:04.736895Z",
+      "number_observed": 1,
+      "objects": {
+        "0": {
+          "type": "ipv4-addr",
+          "value": "192.168.158.5"
+        },
+        "1": {
+          "type": "ipv4-addr",
+          "value": "10.0.0.214"
+        },
+        "2": {
+          "type": "network-traffic",
+          "start": "2020-06-30T19:25:23.518799Z",
+          "end": "2020-06-30T19:27:04.736895Z",
+          "src_ref": "0",
+          "dst_ref": "1",
+          "src_port": 51111,
+          "dst_port": 514,
+          "protocols": [
+            "ip",
+            "tcp"
+          ]
+        },
+        "3": {
+          "type": "user-account",
+          "user_id": "1001",
+          "account_login": "alice"
+        }
+      }
+    },
+    {
+      "type": "observed-data",
+      "id": "observed-data--98e544ee-5085-4ab2-abf0-efc5bf403e6e",
+      "created_by_ref": "identity--89c5ac33-4656-4aa3-be68-f78879e5e273",
+      "created": "2020-06-30T19:31:23.475Z",
+      "modified": "2021-07-14T19:45:16.867Z",
+      "first_observed": "2020-06-30T19:25:23.521241Z",
+      "last_observed": "2020-06-30T19:29:39.826855Z",
+      "number_observed": 1,
+      "objects": {
+        "0": {
+          "type": "ipv4-addr",
+          "value": "192.168.43.225"
+        },
+        "1": {
+          "type": "ipv4-addr",
+          "value": "10.0.0.242"
+        },
+        "2": {
+          "type": "network-traffic",
+          "start": "2020-06-30T19:25:23.521241Z",
+          "end": "2020-06-30T19:29:39.826855Z",
+          "src_ref": "0",
+          "dst_ref": "1",
+          "src_port": 59408,
+          "dst_port": 514,
+          "protocols": [
+            "ip",
+            "tcp"
+          ]
+        },
+        "3": {
+          "type": "user-account",
+          "user_id": "1005",
+          "account_login": "eve"
+        }
+      }
+    },
+    {
+      "type": "observed-data",
+      "id": "observed-data--fbca1584-d7b4-4b41-b408-ec1c90fda398",
+      "created_by_ref": "identity--89c5ac33-4656-4aa3-be68-f78879e5e273",
+      "created": "2020-06-30T19:31:23.316Z",
+      "modified": "2021-07-14T19:45:16.867Z",
+      "first_observed": "2020-06-30T19:25:24.15486Z",
+      "last_observed": "2020-06-30T19:28:51.892722Z",
+      "number_observed": 1,
+      "objects": {
+        "0": {
+          "type": "ipv4-addr",
+          "value": "192.168.107.220"
+        },
+        "1": {
+          "type": "ipv4-addr",
+          "value": "10.0.0.73"
+        },
+        "2": {
+          "type": "network-traffic",
+          "start": "2020-06-30T19:25:24.15486Z",
+          "end": "2020-06-30T19:28:51.892722Z",
+          "src_ref": "0",
+          "dst_ref": "1",
+          "src_port": 61693,
+          "dst_port": 514,
+          "protocols": [
+            "ip",
+            "tcp"
+          ]
+        },
+        "3": {
+          "type": "user-account",
+          "user_id": "1010",
+          "account_login": "juan"
+        }
+      }
+    },
+    {
+      "type": "observed-data",
+      "id": "observed-data--1d7276ba-7f98-4263-8645-b163a69001c6",
+      "created_by_ref": "identity--89c5ac33-4656-4aa3-be68-f78879e5e273",
+      "created": "2020-06-30T19:31:23.523Z",
+      "modified": "2021-07-14T19:45:16.867Z",
+      "first_observed": "2020-06-30T19:25:26.91646Z",
+      "last_observed": "2020-06-30T19:25:27.89832Z",
+      "number_observed": 1,
+      "objects": {
+        "0": {
+          "type": "ipv4-addr",
+          "value": "192.168.107.220"
+        },
+        "1": {
+          "type": "ipv4-addr",
+          "value": "10.0.0.134"
+        },
+        "2": {
+          "type": "network-traffic",
+          "start": "2020-06-30T19:25:26.91646Z",
+          "end": "2020-06-30T19:25:27.89832Z",
+          "src_ref": "0",
+          "dst_ref": "1",
+          "src_port": 55606,
+          "dst_port": 22,
+          "protocols": [
+            "ip",
+            "tcp"
+          ]
+        },
+        "3": {
+          "type": "user-account",
+          "user_id": "1005",
+          "account_login": "eve"
+        }
+      }
+    },
+    {
+      "type": "observed-data",
+      "id": "observed-data--67fc2b3a-d764-4357-8fc4-f1e30c4795e2",
+      "created_by_ref": "identity--89c5ac33-4656-4aa3-be68-f78879e5e273",
+      "created": "2020-06-30T19:31:23.495Z",
+      "modified": "2021-07-14T19:45:16.867Z",
+      "first_observed": "2020-06-30T19:25:26.950946Z",
+      "last_observed": "2020-06-30T19:27:19.969637Z",
+      "number_observed": 1,
+      "objects": {
+        "0": {
+          "type": "ipv4-addr",
+          "value": "192.168.219.106"
+        },
+        "1": {
+          "type": "ipv4-addr",
+          "value": "10.0.0.134"
+        },
+        "2": {
+          "type": "network-traffic",
+          "start": "2020-06-30T19:25:26.950946Z",
+          "end": "2020-06-30T19:27:19.969637Z",
+          "src_ref": "0",
+          "dst_ref": "1",
+          "src_port": 58012,
+          "dst_port": 22,
+          "protocols": [
+            "ip",
+            "tcp"
+          ]
+        },
+        "3": {
+          "type": "user-account",
+          "user_id": "1013",
+          "account_login": "maria"
+        }
+      }
+    },
+    {
+      "type": "observed-data",
+      "id": "observed-data--a871fdd3-2a8d-48c4-9e18-94799ea4e833",
+      "created_by_ref": "identity--89c5ac33-4656-4aa3-be68-f78879e5e273",
+      "created": "2020-06-30T19:31:23.491Z",
+      "modified": "2021-07-14T19:45:16.867Z",
+      "first_observed": "2020-06-30T19:25:27.798203Z",
+      "last_observed": "2020-06-30T19:26:11.871085Z",
+      "number_observed": 1,
+      "objects": {
+        "0": {
+          "type": "ipv4-addr",
+          "value": "192.168.157.211"
+        },
+        "1": {
+          "type": "ipv4-addr",
+          "value": "10.0.0.201"
+        },
+        "2": {
+          "type": "network-traffic",
+          "start": "2020-06-30T19:25:27.798203Z",
+          "end": "2020-06-30T19:26:11.871085Z",
+          "src_ref": "0",
+          "dst_ref": "1",
+          "src_port": 52143,
+          "dst_port": 22,
+          "protocols": [
+            "ip",
+            "tcp"
+          ]
+        },
+        "3": {
+          "type": "user-account",
+          "user_id": "1004",
+          "account_login": "david"
+        }
+      }
+    }
+  ]
+}

--- a/tests/test_storage.py
+++ b/tests/test_storage.py
@@ -278,15 +278,16 @@ def test_reassign(fake_bundle_file, fake_csv_file, tmpdir):
     for url in urls:
         url['x_enrich'] = 1
 
-    # Now reload into a new var
-    store.reassign('enriched_urls', urls)
+    # Now reload into the same var
+    store.reassign('urls', urls)
     rows = store.lookup('__membership')
     print(json.dumps(rows, indent=4))
-    rows = store.lookup('enriched_urls')
+    rows = store.lookup('urls')
     print(json.dumps(rows, indent=4))
     assert len(rows) == len(urls)
+    assert rows[0]['x_enrich'] == 1
 
-    # Make sure original var isn't modified
+    # Make sure original var length isn't modified
     urls = store.lookup('urls')
     assert len(urls) == 14
 
@@ -304,7 +305,7 @@ def test_appdata(fake_bundle_file, tmpdir):
     assert data['foo'] == result['foo']
     assert len(result) == len(data)
 
-    store2 = tmp_storage(tmpdir)
+    store2 = tmp_storage(tmpdir, clear=False)
     result = json.loads(store2.get_appdata('ssh_conns'))
     assert data['foo'] == result['foo']
     assert len(result) == len(data)
@@ -378,7 +379,6 @@ def test_merge(fake_bundle_file, tmpdir):
     assert merged == all_urls
 
 
-@pytest.mark.skip(reason="this only happens with postgresql")
 def test_change_type(fake_bundle_file, tmpdir):
     store = tmp_storage(tmpdir)
     store.cache('q1', [fake_bundle_file])
@@ -392,9 +392,7 @@ def test_change_type(fake_bundle_file, tmpdir):
     # Create a var `sorted_foo` of type url that depends on `foo`
     store.assign('sorted_foo', 'foo', op='sort', by='value')
 
-    # sqlite3 doesn't have issues with this; only PostgreSQL
-    with pytest.raises(IncompatibleType):
-        store.extract('foo', 'ipv4-addr', 'q1', "[ipv4-addr:value ISSUBSET '192.168.0.0/16']")
+    store.extract('foo', 'ipv4-addr', 'q1', "[ipv4-addr:value ISSUBSET '192.168.0.0/16']")
 
 
 def test_remove(fake_bundle_file, tmpdir):
@@ -462,3 +460,16 @@ def test_remove_after_merge(fake_bundle_file, tmpdir, names):
 
     merged = set(store.values('url:value', 'merged'))
     assert merged == all_urls
+
+
+def test_port_zero(fake_bundle_file_2, tmpdir):
+    store = tmp_storage(tmpdir)
+    store.cache('q1', [fake_bundle_file_2])
+
+    store.extract('conns', 'network-traffic', 'q1', "[network-traffic:dst_port < 1024]")
+
+    # sort by src_port and make sure port 0 comes first
+    store.assign('sconns', 'conns', op='sort', by='src_port')
+    conns = store.lookup('sconns')
+    assert conns[0]['src_port'] == 0
+    assert conns[0]['id'] == 'network-traffic--2171d844-d635-4f03-91cc-0a36f1caf3b6_2'

